### PR TITLE
[KOA-5077] BPKDialog SwiftUI

### PR DESCRIPTION
**Pull request by frugoman**
+ [ ] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-ios/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [ ] `README.md`
+ [ ] Tests
+ [ ] [Screenshotting code](https://github.com/Skyscanner/backpack-ios/blob/main/Example/Backpack%20Screenshot/Screenshots.swift)
+ [ ] Adding a component? Remember to expose it in the [main `Backpack.h` header file](https://github.com/Skyscanner/backpack-ios/tree/main/Backpack/Backpack.h)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: Backpack-SwiftUI/Dialog/Classes/BPKDialog.swift [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+public extension View {
+    /// Displays a success dialog with a title, text and a list of buttons.
+    func bpkSuccessDialog(
+        presented: Binding<Bool>,
+        title: String,
+        text: String,
+        confirmButton: BPKDialogButton,
+        secondaryButton: BPKDialogButton? = nil,
+        linkButton: BPKDialogButton? = nil,
```
Comment (marianeum) [inline Backpack-SwiftUI/Dialog/Classes/BPKDialog.swift:29]: when talking to adam the link button should only be used when the secondary button is also available. on compose we therefore created another function for this which makes the secondary button required if link is provided

File: Backpack-SwiftUI/Dialog/Classes/BPKDialog.swift [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        var buttons = [BPKDialogButton(button: confirmButton, style: .featured)]
+        if let secondaryButton {
+            buttons.append(BPKDialogButton(button: secondaryButton, style: .secondary))
+        }
+        if let linkButton {
+            buttons.append(BPKDialogButton(button: linkButton, style: .link))
+        }
+        return self.modifier(DialogContainerViewModifier(isPresented: presented, dialogContent: {
+            DialogWithIconContent(
+                icon: .tick,
```
Comment (marianeum) [inline Backpack-SwiftUI/Dialog/Classes/BPKDialog.swift:41]: the icon should be configurable (the colour not)

File: Backpack-SwiftUI/Dialog/Classes/DialogTextContent.swift [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        self.contentAlignment = contentAlignment
+    }
+    
+    var body: some View {
+        VStack(alignment: contentAlignment, spacing: spacing) {
+            HStack {
+                if contentAlignment == .trailing {
+                    Spacer()
+                }
+                BPKText(title, style: .heading3)
```
Comment (gert-janvercauteren) [inline Backpack-SwiftUI/Dialog/Classes/DialogTextContent.swift:46]: add `.accessibilityAddTraits(.isHeader)` to mark this as the heading of the dialog.

**Review by gert-janvercauteren — COMMENTED** [review]
Left some accessibility improvements.
We also need to make sure the scrim is 'focusable' to dismiss the popup. 
And (if cancelable, make sure we have the perform escape enabled (draw Z))

File: Backpack-SwiftUI/Dialog/Classes/DialogWithIconContent.swift [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            VStack(spacing: BPKSpacing.lg.value) {
+                textContent
+                actions
+            }
+            .padding(.lg)
+            .padding(.top, 12)
+            .background(.surfaceDefaultColor)
+            .clipShape(RoundedRectangle(cornerRadius: .md))
+            .shadow(.lg)
+        }
```
[the remark this answers is not in the document]
Comment (frugoman) [reply Backpack-SwiftUI/Dialog/Classes/DialogWithIconContent.swift:41]: This was solved in a different way (as the modal now is presented using a UIViewController)

File: Backpack-SwiftUI/Dialog/Classes/BPKDialog.swift (hunk 1 above)
Comment (frugoman) [reply Backpack-SwiftUI/Dialog/Classes/BPKDialog.swift:29]: done

File: Backpack-SwiftUI/Dialog/Classes/BPKDialog.swift (hunk 2 above)
Comment (frugoman) [reply Backpack-SwiftUI/Dialog/Classes/BPKDialog.swift:41]: done

**Comment by frugoman** [comment]
> Left some accessibility improvements. We also need to make sure the scrim is 'focusable' to dismiss the popup. And (if cancelable, make sure we have the perform escape enabled (draw Z))

I couldnt do the Z gesture to try it out, please feel free to try it out!
I'm using a UIViewController to actually present the dialog, as all other alternatives lacked something or were straight broken!
As it's a modally presented view controller, the Z gesture should work out of the box? (if it something all modals have on ios)

I checked out the native Alert, and that does not support selecting the scrim at all, so I did not implement this here. but if you still think is a good idea to have I could give that a try! (I think it would be with the SortPiority property, but not sure what traits to give the scrim, seems more hacky than just not supporting it)

**Review by gert-janvercauteren — DISMISSED** [review]
Dialog now behaves as expected with VoiceOver 👏 

Only things is to match the documentation screenshots with Compose and UIKit. 
https://www.skyscanner.design/latest/components/dialog/i-os.html

Then this component is ready to ship 🚢

File: Backpack-SwiftUI/Dialog/Classes/DialogContainerViewModifier.swift [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                }
+            }
+    }
+    
+    private func showDialogWithContent<Content: View>(_ content: () -> Content) {
+        let controller = UIHostingController(rootView: content())
+        controller.view.backgroundColor = .clear
+        controller.modalTransitionStyle = .crossDissolve
+        controller.modalPresentationStyle = .overFullScreen
+        UIApplication.shared.keyWindow?.rootViewController?.present(controller, animated: true)
```
Comment (gert-janvercauteren) [inline Backpack-SwiftUI/Dialog/Classes/DialogContainerViewModifier.swift:63]: We should not use this as we will support multiple windows in the future.
https://stackoverflow.com/questions/57134259/how-to-resolve-keywindow-was-deprecated-in-ios-13-0

**Review by gert-janvercauteren — COMMENTED** [review]
Waiting for updated docs screenshots

**Review by marianeum — APPROVED** [review]

**Pull request merged**